PyExpLabSys/drivers/mks_g_series.py

```diff
@@ -20,46 +36,57 @@ def __init__(self, port='/dev/ttyUSB0'):
         self.ser.bytesize = serial.EIGHTBITS
         self.ser.stopbits = serial.STOPBITS_ONE
 
-    def checksum(self, command, reply=False):
+    def checksum(self, command):
         """ Calculate checksum of command """
-        if not reply:
-            com_string = '@' + command
-        else:
-            com_string = command
-        total = 0
-        for i in range(0, len(com_string)):
-            total = total + ord(com_string[i])
-        return (hex(total)[-2:]).upper()
+        com_string = '@' + command
+        checksum = sum([ord(i) for i in com_string])
+        # The following hack deviates from the manual, but has proven to work on two different setups
+        if 'NAK' in com_string:
+            checksum += 9
+        return hex(checksum)[-2:].upper()
 
     def comm(self, command, addr):
         """ Implements communication protocol """
+        # TODO: add retries on comm errors
         com_string = str(addr).zfill(3) + command + ';'
         checksum = self.checksum(com_string)
-        com_string = '@@@@' + com_string + checksum
+        com_string = '@@@' + com_string + checksum
         com_string = com_string.encode('ascii')
         self.ser.write(com_string)
         time.sleep(0.1)
-        reply = self.ser.read(self.ser.inWaiting())
-        try:
-            reply = reply.decode('ascii')
-        except UnicodeDecodeError:
-            reply = reply.decode('ascii', 'ignore')
-            reply = reply.strip('\x00')
-            reply = '@' + reply
-
+        raw_reply = self.ser.read(self.ser.inWaiting())
+        # In case that noise adds zeros on either side:
+        reply = raw_reply.strip(b'\x00')
+        reply = reply.decode('ascii')
         if len(reply) == 0:
             LOGGER.warning('No such device')
-        else:
-            if reply[-3:] == self.checksum(reply[1:-3], reply=True):
-                reply = reply[6:-3]  # Cut away master address and checksum
+            return ''
+        # Response structure: '@@@000' + 'ACK' + content + ';' + 2-char-checksum
+        noise_check_passed = reply.find(';') == len(reply) - 3
+        ack = reply[6:9]
+        content = reply[9:-3]
+        crc = reply[-2:]
+        if noise_check_passed:
```

Comment:
I find the triple nested if-else here a bit of a challenge to read. How about flipping the `noise_check_passed` check around and do an early return. That will allow you to dedent once:

```python
    if not noise_check_passed:
        LOGGER.warning('Response seemingly too noisy: {}'.format(repr(raw_reply)))
        return ''

    if crc == self.checksum(reply[1:-2]):  # NOW AT METHOD BASE INDENT
        ...
```
and hell, while we are at it, you could take it one step further and reverse the crc checksum as well and do another early return there, you can dedent once more:

```python
    if not noise_check_passed:
        LOGGER.warning('Response seemingly too noisy: {}'.format(repr(raw_reply)))
        return ''

    if crc != self.checksum(reply[1:-2]):
        LOGGER.error('Checksum error in reply')
        return ''

    if ack == 'ACK':
        return content
```

The effect is to flatten the whole thing, which I think most find easier to read. The other advantage is that it lessens the amount of stuff to keep in ones head. With a structure like this, it will have the effect of a sort of filter, which is easier to reason about. "If it's end-noisy, give up and return. If the check fails, give and and return, and then, when all is good with the structure, parse the stuff.

Reply:
Excellent comment! I have implemented it in the recent commit.
I think it was last minute I changed all the `if True: reply = X; elif Actually_True: reply = Y; else: reply = ''; return reply` and as long as I had the mixed structure, the deep nesting was to some degree necessary and so I didn't stop to consider it again afterwards.. :)